frontend/test/pytest/test_autograph.py

```diff
@@ -432,5 +506,327 @@ def circuit():
             qjit(autograph=True)(qml.qnode(qml.device(backend, wires=1))(circuit))
 
 
+class TestForLoops:
```

Comment:
These are really good integration tests @dime10 👍 

It might make sense though to also add more unit-test style tests. That is, tests that, for a given circuit, compares the autograph output to the expected output.

I'm not sure exactly how this could be done, but perhaps the `autograph_source` function could be used?

The main reason being that integration tests, since they're including circuit simulation, might not always represent accurate conversion (e.g., simple circuits could return the same two values for different structures, potentially because of periodicity, or because of bugs/gates not affecting the state). This is something that has happened a lot in the standard PL test suite over the years.

Reply:
Hmm yes, we do have unit tests checking the autograph source code transformation over in the [lit tests](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/test/lit/test_autograph.py). I will add something there for the for loop as well, although the testing is less useful than I would like because autograph will just generate `__ag.if_stmt`s or `__ag.for_stmt`s, rather than generating the Catalyst `cond` or `for_loop` directly.

Reply:
So I've ended up adding a bunch of tests which are more unit-testy and which actually uncovered some bugs: https://github.com/PennyLaneAI/catalyst/pull/258/commits/dfd98a7eb296cd6517cadc7ef14a7ab738a2a438
In the process I also improved the error checking:
https://github.com/PennyLaneAI/catalyst/pull/258/commits/8d2a40ce606d7d1206ddecd67924783da047354a

Let me know if you think this is sufficient :)

Reply:
Nice, I love to hear that 💪